src/prefect/server/database/orm_models.py

```diff
@@ -137,86 +144,77 @@ def __tablename__(cls):
     )
 
-    flow_runs = sa.orm.relationship("FlowRun", back_populates="flow", lazy="raise")
-    deployments = sa.orm.relationship("Deployment", back_populates="flow", lazy="raise")
-
-    __table_args__ = (
+    __table_args__: Any = (
         sa.UniqueConstraint("name"),
         sa.Index("ix_flow__created", "created"),
+        sa.Index("trgm_ix_flow_name", "name", postgresql_using="gin").ddl_if(
+            dialect="postgresql"
+        ),
     )
 
 
 class FlowRunState(Base):
     """SQLAlchemy mixin of a flow run state."""
 
-    flow_run_id = sa.Column(
-        UUID(), sa.ForeignKey("flow_run.id", ondelete="cascade"), nullable=False
+    flow_run_id: Mapped[uuid.UUID] = mapped_column(
+        sa.ForeignKey("flow_run.id", ondelete="cascade")
     )
 
-    type = sa.Column(
-        sa.Enum(schemas.states.StateType, name="state_type"), nullable=False, index=True
+    type: Mapped[schemas.states.StateType] = mapped_column(
+        sa.Enum(schemas.states.StateType, name="state_type"), index=True
     )
-    timestamp = sa.Column(
-        Timestamp(),
-        nullable=False,
-        server_default=now(),
-        default=lambda: pendulum.now("UTC"),
+    timestamp: Mapped[pendulum.DateTime] = mapped_column(
+        server_default=now(), default=lambda: pendulum.now("UTC")
     )
-    name = sa.Column(sa.String, nullable=False, index=True)
-    message = sa.Column(sa.String)
-    state_details = sa.Column(
+    name: Mapped[str] = mapped_column(index=True)
+    message: Mapped[Optional[str]]
+    state_details: Mapped[schemas.states.StateDetails] = mapped_column(
         Pydantic(schemas.states.StateDetails),
         server_default="{}",
         default=schemas.states.StateDetails,
-        nullable=False,
     )
-    _data = sa.Column(sa.JSON, nullable=True, name="data")
+    _data: Mapped[Optional[Any]] = mapped_column(JSON, name="data")
 
-    result_artifact_id = sa.Column(
-        UUID(),
-        sa.ForeignKey(
-            "artifact.id",
-            ondelete="SET NULL",
-            use_alter=True,
-        ),
+    result_artifact_id: Mapped[Optional[uuid.UUID]] = mapped_column(
+        sa.ForeignKey("artifact.id", ondelete="SET NULL", use_alter=True),
         index=True,
     )
 
-    _result_artifact = sa.orm.relationship(
-        "Artifact",
+    _result_artifact: Mapped[Optional["Artifact"]] = relationship(
         lazy="selectin",
         foreign_keys=[result_artifact_id],
         primaryjoin="Artifact.id==FlowRunState.result_artifact_id",
     )
 
     @hybrid_property
-    def data(self):
+    def data(self) -> Optional[Any]:
         if self._data:
             # ensures backwards compatibility for results stored on state objects
             return self._data
         if not self.result_artifact_id:
             # do not try to load the relationship if there's no artifact id
             return None
+        if TYPE_CHECKING:
+            assert self._result_artifact is not None
```

Comment:
hi @mjpieters - I know this PR is stacked on the other one, just curious about why this is necessary

Reply:
When validating the code with a type checker, the checker rightly points out that `self._result_artifact` could be `None`. The type checker can't know that when `self.result_artifact_id` is not `None`, then `self._result_artifact` will also exist. The `assert` acts as a type guard here, telling the type checker to trust that `self._result_artifact` is not `None` here.

The `if TYPE_CHECKING:` block makes it that the type guard is never actually executed in production code; the Python byte compiler removes the whole block from the compiled bytecode.